test/unittests/t_catalog_mgr.cc

```diff
@@ -0,0 +1,174 @@
+ * This file is part of the CernVM File System.
+ */
+
+#include <gtest/gtest.h>
+
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include "../../cvmfs/catalog.h"
+#include "../../cvmfs/catalog_mgr.h"
+#include "../../cvmfs/hash.h"
+#include "../../cvmfs/shortstring.h"
+#include "../../cvmfs/util.h"
+#include "testutil.h"
+
+using namespace std;  // NOLINT
+
+
+namespace catalog {
+
+class T_CatalogManager : public ::testing::Test {
+ protected:
+  void SetUp() {
+    catalog_mgr_ = new MockCatalogManager(&statistics_);
+  }
+
+  void TearDown() {
+    delete catalog_mgr_;
+  }
+
+  void AddTree() {
+    const shash::Any empty_hash = shash::Any();
+    const int size = 32;
+    const size_t file_size = 4096;
+    const char suffix = shash::kSha1;
+    shash::Any hash;
+    shash::Any empty_content;
+    // adding ""
+    MockCatalog *root_catalog = catalog_mgr_->RetrieveRootCatalog();
+    root_catalog->AddFile(empty_content, file_size, "", "");
+    // adding "/dir"
+    root_catalog->AddFile(empty_content, file_size, "", "dir");
+    // adding "/file1"
+    hash = shash::Any(shash::kSha1,
+                      reinterpret_cast<const unsigned char*>(hashes[1]),
+                      size, suffix);
+    root_catalog->AddFile(hash, file_size, "", "file1");
+    // adding "/dir/dir"
+    root_catalog->AddFile(empty_content, file_size, "/dir", "dir");
+    // adding "/dir/dir/file2"
+    hash = shash::Any(shash::kSha1,
+                      reinterpret_cast<const unsigned char*>(hashes[1]),
+                      size, suffix);
+    root_catalog->AddFile(hash, file_size, "/dir/dir", "file2");
+    // adding "/dir/dir/dir"
+    root_catalog->AddFile(empty_content, file_size, "/dir/dir", "dir");
+    // adding a nested catalog in "/dir/dir/dir"
+    string catalog_path_str = "/dir/dir/dir";
+    PathString catalog_path(catalog_path_str.c_str(),
+                            catalog_path_str.length());
+    MockCatalog *new_catalog = new MockCatalog("/dir/dir/dir", shash::Any(),
+                                               4096, 1, 0, false,
+                                               root_catalog, NULL);
+    ASSERT_NE(static_cast<MockCatalog*>(NULL), new_catalog);
+    ASSERT_EQ(1u, root_catalog->GetChildren().size());
+    catalog_mgr_->RegisterNewCatalog(new_catalog);
+    // adding "/dir/dir/dir/file3" to the new nested catalog
+    hash = shash::Any(shash::kSha1,
+                          reinterpret_cast<const unsigned char*>(hashes[2]),
+                          size, suffix);
+    new_catalog->AddFile(hash, file_size, "/dir/dir/dir", "file3");
+    // adding "/dir/dir/dir/file4" to the new nested catalog
+    hash = shash::Any(shash::kSha1,
+                          reinterpret_cast<const unsigned char*>(hashes[3]),
+                          size, suffix);
+    new_catalog->AddFile(hash, file_size, "/dir/dir/dir", "file4");
+    // we haven't mounted the second catalog yet!
+    ASSERT_EQ(1, catalog_mgr_->GetNumCatalogs());
+  }
```

Comment:
Please add a couple of more catalogs to make the test a bit more realistic. For example, a single lookup should cause a whole branch of catalogs to be loaded. Like:
```bash
ls /dir/dir/dir/dir/dir/dir
```
Should load **/dir/dir** and in the same lookup **/dir/dir/dir/dir**.